File: leetcode/binary_search/3506_Find_Time_Required_to_Eliminate_Bacterial_Strains/codes/heap.py
from typing import List
import heapq

class Solution:
    def minEliminationTime(self, timeReq: List[int], splitTime: int) -> int:
        heapq.heapify(timeReq)

        while len(timeReq) > 1:
            t1, t2 = heapq.heappop(timeReq), heapq.heappop(timeReq)
            t2 += splitTime
            heapq.heappush(timeReq, t2) 

        return timeReq[0]

File: leetcode/binary_search/3506_Find_Time_Required_to_Eliminate_Bacterial_Strains/codes/heap.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    long long minEliminationTime(vector<int>& timeReq, int splitTime) {
        priority_queue<long long, vector<long long>, greater<long long>> pq;

        for (int t : timeReq)
            pq.push(t);

        while (pq.size() > 1) {
            long long t1 = pq.top(); pq.pop();
            long long t2 = pq.top(); pq.pop();

            long long merged = t2 + splitTime;
            pq.push(merged);
        }

        return pq.top();
    }
};
